core/unit_test/TestMathematicalFunctions.hpp

```diff
@@ -0,0 +1,831 @@
+template <> struct math_binary_function_return_type<             float,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<            double,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,              float> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,             double> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,               bool> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,              short> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,                int> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,               long> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,          long long> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,     unsigned short> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,       unsigned int> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double,      unsigned long> { using type = long double; };
+template <> struct math_binary_function_return_type<       long double, unsigned long long> { using type = long double; };
+template <> struct math_binary_function_return_type<             short,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<               int,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<              long,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<         long long,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<    unsigned short,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<      unsigned int,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<     unsigned long,        long double> { using type = long double; };
+template <> struct math_binary_function_return_type<unsigned long long,        long double> { using type = long double; };
+#endif
+template <class T, class U>
+using math_binary_function_return_type_t = typename math_binary_function_return_type<T, U>::type;
+// clang-format on
+
+struct FloatingPointComparison {
+  double rel_tol_ = 1e-6;
+  double abs_tol_ = 1e-12;
```

Comment:
shouldn't this be different for flaot and double as return type?

Reply:
They are casted in the fp comparison.  Do you see an issue with it?

Reply:
resolved by my commit. Now we are using FLT_EPSILON and DBL_EPSILON combined with a ULP_factor which depends on the operation.